### DISCUSSION TITLE
UI library for React project and long term plan

### DISCUSSION THREAD

**Discussion by MCatherine1994** [bcgov/nr-theme#38] [Ideas]
Based on our recent discussion, we had some concerns with the Carbon UI library:

- Not support Vue3 
- Need manually change to make components responsive
- Some documentation is hard to find and follow
- Restrict component features

Previous thoughts of using Carbon for React project in FSA (when we already started SPAR):

- Get support from Encora since they already use Carbon React for SPAR, can reuse their existing design token code and component override stylesheet
- Have existing components in Figma that was built with Carbon Design System, can benefit from using the Carbon UI library, and do less custom styling changes

Advantages of using Carbon:

- Good icons (if we use another library, can we still get the icons from them?)
- Multiple size components for different screen sizes
- Dark mode theme

Next step for React project (Jas & Craig):

- What UI library want to use, it's better to pick the same one for SPAR and RESULT/SILVA, otherwise we still need to maintain more stylesheets

Next step for common theme repo:

- If we can share the same stylesheet between different React/Vue3 project? Will that require the same UI library? Or we can just let teams use the classname, and just control the basic style
- Consider the PrimeReact library

Next step for service designers and developers to discuss:

- What we want to control at the team level? Based on the different requirement for each team, it's very possible they need to use other libraries to support their own needs, it's hard for us to control everything. I would suggest we only maintain simple components with basic style, for complex ones, allow each team to take care of
- How much difference we can accept between the design in Figma and what we actually present in the project? If we'll use different UI library, there will be difference between the design in Figma and the real component. It will be great if we can make changes in Figma as well, for the ones we all agree for the difference
- How we keep track if our stylesheet is up to date with what in the Figma? For long term consideration, it's better that Figma can have some version control as well. So our common stylesheet can tell which Figma design version it refers to. Otherwise in the future when we make more changes in the design system, we'll loose track

**Comment by mamartinezmejia** [comment]
I would say that Carbon has some limitations compared to Bootstrap and PrimeFaces (PrimeVue or PrimeReact). Here some of them:

"
1. Complexity: The Carbon Design System is a comprehensive and feature-rich UI framework, which can make it more complex to learn and implement compared to simpler frameworks or libraries. It may have a steeper learning curve, especially for developers who are new to it or have limited experience with similar frameworks.

2. Overhead: Because of its extensive feature set, Carbon Design System might introduce some overhead in terms of file size and performance. This can impact the loading time of your web or application if not optimized properly. It's crucial to consider the specific requirements and constraints of your project and ensure that the framework aligns with your performance goals.

3. Customization limitations: While Carbon Design System offers a wide range of pre-designed components and styles, it may have certain limitations when it comes to customization. Depending on your project's requirements, you may find it challenging to deviate from the provided design patterns and styles. Customizing the framework extensively or achieving a unique look and feel might require additional effort or modifications to the core components.

4. Compatibility: The Carbon Design System might not be compatible with all browsers or devices, particularly older ones. It's essential to review the framework's documentation to ensure compatibility with your target audience's browsers and devices. Additionally, updates to browsers or devices could potentially introduce compatibility issues, so it's important to stay up to date with the framework's releases and ensure compatibility with the latest software versions.

5. Community and ecosystem: While the Carbon Design System has gained popularity and adoption, its community and ecosystem might be relatively smaller compared to some other UI frameworks like Bootstrap or Material-UI. This can impact the availability of community-contributed resources, third-party integrations, and support. However, this disadvantage might have changed over time as the framework gained more traction."

**Reply by jackwingsam** [reply]
For points 1 to 4, would these not apply to any design system? Are you advocating that we don't use a design system at all?

**Reply by mamartinezmejia** [reply]
Hi Jack,

1. Regarding your 1st question, the points I wrote are downsides or limitations about Carbon. Bootstrap, for example, is a widely used and known for its simplicity and responsiveness, it is well known for being responsive and it is compatible with many browsers and devices. 

So, no, it doesn't apply to any design system.


2. Regarding your 2nd question, I don't understand why you are asking that, as I am not suggesting we shouldn't use a design system at all. I simply presented some points that have been raised by developer communities.

Currently, we are using Bootstrap as our framework. The look and feel that we all want to achieve (Figma) can be customized for **any** UI library. It's just a matter of playing with the CSS and making it look very close to what Carbon offers, using the Figma style that is partially used by SPAR.

So, no, I am not advocating against using a design system.

**Comment by jackwingsam** [comment]
In order for us to move away from Carbon, we need to demonstrate that whatever new system we pick can replicate what we are currently using from Carbon. No empty promises about infinite possibilities please. We need to see production level examples because SPAR is already using Carbon. 

In my mind, we need to see at a minimum: (This is not an exhaustive list)

1. SPAR's icons replicated
2. SPAR's components replicated
3. Dark mode colours replicated

With that work in mind, we then we have to evaluate how much work it would actually take to move off Carbon. Again, because SPAR is already using it. Then all the different FSA teams have to understand and absorb that level of work involved to move off Carbon. 

If it costs more time and effort to move from Carbon to another system, then it will not be a valuable move even if Carbon is not ideal.
